data/cube/impl/BaseRow.js

```diff
@@ -0,0 +1,151 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {has, isEmpty, reduce} from 'lodash';
+
+
+/**
+ * Base class for a view row.
+ */
+export class BaseRow {
+
+    view = null;
+    id = null;
+    parent = null;
+    children = null;
+    data;
+
+    get isLeaf()        {return false}
+    get isAggregate()   {return false}
+    get isBucket()      {return false}
+
+    constructor(view, id) {
+        this.view = view;
+        this.id = id;
+        this.data = {id, _meta: this};
+    }
+
+    //-----------------------
+    // For all rows types
+    //------------------------
+    noteBucketed(bucketSpec, bucketVal) {
+        this.data.buckets = this.data.buckets ?? {};
+        this.data.buckets[bucketSpec.name] = bucketVal;
+        this.children?.forEach(it => it.noteBucketed(bucketSpec, bucketVal));
+    }
+
+    // Determine what should be exposed as the actual children in the
+    // row data.  This where we lock, skip degenerate rows, etc.
+    applyVisibleChildren() {
+        let {children, view, data} = this,
+            {lockFn, omitFn} = view.cube;
+
+        if (!children) return null;
+
+        // Remove all children from the data if the query is not configure to include leaves and
+        // this row has leaves as children
+        if (!view.query.includeLeaves && children[0]?.isLeaf) {
+            data.children = null;
+            return;
+        }
+
+        // Check if we need to 'lock' this row - removing all children from the data so they will not
+        // appear in the UI but remain as children of this AggregateMeta to ensure that updates to
+        // those child rows will update our aggregate value
+        if (lockFn && lockFn(this)) {
+            this.locked = true;
+            data.children = null;
+            return;
+        }
+
+        // Apply recursively -- we need to go depth first to allow recursive collapsing
+        children.forEach(it => it.applyVisibleChildren());
+
+        // Skip cullable single children, by wiring up to *their* data children.
+        if (children.length === 1) {
+            const childRow = children[0];
+            if (this.isRedundantChild(childRow) || (omitFn && omitFn(childRow))) {
+                data.children = childRow.data.children;
+                return;
+            }
+        }
+
+        // ...otherwise wire up to your own children's data
+        data.children = children.map(it => it.data);
+    }
+
+    isRedundantChild(row) {
+        // TODO:  put this test in application code omitFn instead?
```

Comment:
I could see this being an option on the Query - `cullRedundantChildren` or something

Reply:
yep -- let's see how things shake out